### PR TITLE
No Longer Decoding URL-Encoded Blob Name and only Storing Original Name

### LINKED ISSUE #37564: [BUG] Difference in URI handling behavior of % for Blob Storage between V8 and V12
**Describe the bug**
When trying to upgrade between V8 and V12 I have discovered differing behavior when handling % in a file path in blob storage.

On v8 the code looks like this:
```
  CloudBlobContainer container = getOrCreateCloudBlobContainer(containerName); // create a CloudBlobContainer
  try (FileInputStream stream = new FileInputStream(file)) {
    container.getBlockBlobReference(blobPath).upload(stream, file.length());
  }
```

On v12 the code looks like this:
```
    BlobContainerClient blobContainerClient = getOrCreateBlobContainerClient(containerName);

    try (FileInputStream stream = new FileInputStream(file)) {
      // By default this creates a Block blob, no need to use a specific Block blob client.
      blobContainerClient.getBlobClient(blobPath).upload(stream, file.length());
    }
```

Where blobPath is "ab2a7d5f-b973-4222-83ba-d0581817a819 %Россия 한국 中国!?/file" (anything with a %<something after it> works to trigger this issue)

The public docs say that the client should automatically handle url encoding (% is not in the reserved characters list) (https://learn.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata).

But when I upload a file with a unencoded special character name directly via the azure UI it seems to work okay.
The V8 client seems to follow this logic, as it will accept a unencoded special character name.

But the V12 client will throw an exception.
***Exception or Stack Trace***
```
[2023-11-08T16:46:06.435Z] Caused by: java.lang.IllegalArgumentException: URLDecoder: Illegal hex characters in escape (%) pattern - Error at index 0 in: "Ро"
[2023-11-08T16:46:06.435Z] 	at java.base/java.net.URLDecoder.decode(URLDecoder.java:232)
[2023-11-08T16:46:06.435Z] 	at java.base/java.net.URLDecoder.decode(URLDecoder.java:142)
[2023-11-08T16:46:06.435Z] 	at com.azure.storage.common.Utility.decode(Utility.java:88)
[2023-11-08T16:46:06.435Z] 	at com.azure.storage.common.Utility.urlDecode(Utility.java:55)
[2023-11-08T16:46:06.435Z] 	at com.azure.storage.blob.specialized.BlobAsyncClientBase.<init>(BlobAsyncClientBase.java:242)
[2023-11-08T16:46:06.435Z] 	at com.azure.storage.blob.specialized.BlobAsyncClientBase.<init>(BlobAsyncClientBase.java:206)
[2023-11-08T16:46:06.435Z] 	at com.azure.storage.blob.BlobAsyncClient.<init>(BlobAsyncClient.java:154)
[2023-11-08T16:46:06.435Z] 	at com.azure.storage.blob.BlobContainerAsyncClient.getBlobAsyncClient(BlobContainerAsyncClient.java:194)
[2023-11-08T16:46:06.435Z] 	at com.azure.storage.blob.BlobContainerAsyncClient.getBlobAsyncClient(BlobContainerAsyncClient.java:172)
[2023-11-08T16:46:06.435Z] 	at com.azure.storage.blob.BlobContainerClient.getBlobClient(BlobContainerClient.java:98)
```

**To Reproduce**
Try using the Java V12 blob client to upload a file with blobPath: "asdf%asdf"

***Code Snippet***
Linked these in the description

**Expected behavior**
I would expect the new client's behavior to be similar to both the docs and the UI, but right now the docs and the UI have different behavior.

**Screenshots**
N/A

**Setup (please complete the following information):**
 - OS: Ubuntu
 - IDE: N/A
 - Library/Libraries: azure-storage-blob:12.24.0
 - Java version: 8,11,17
 - App Server/Environment: N/A
 - Frameworks: N/A

**Additional context**

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [X] Bug Description Added
- [X] Repro Steps Added
- [X] Setup information Added

### PR DESCRIPTION
**Pull request by ibrahimrabab**
# Description

resolves #37564

Previously, when calling `BlobContainerClient.getBlobClient(String blobName)` or `BlobClient.getBlobName()` with a URL-encoded blob name, the blob name would be decoded and stored and uploaded as the decoded name. Now when passing in a URL-encoded blob name, it will no longer be decoded. This is to ensure that the blob name is not modified in any way. If blob name needs to be decoded, use `URLDecoder.decode()` or pass the exact name of the blob to upload into `BlobContainerClient.getBlobClient(String blobName)`. 

This change follows the same pattern as done in .NET and Python.

### CODE REVIEW

File: sdk/storage/azure-storage-blob/CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,10 @@
 ### Features Added
 
 ### Breaking Changes
+- When calling `BlobContainerClient.getBlobClient(String blobName)` or `BlobClient.getBlobName()` with a URL-encoded 
+blob name, the blob name will no longer be decoded. This is to ensure that the blob name is not modified in any way. If
```
Comment (seanmcc-msft) [inline sdk/storage/azure-storage-blob/CHANGELOG.md:9]: I think this changlog entry is ambiguous and could confuse customers.  We should explain the new logic the way I did in my email about this feature.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobApiTests.java [hunk 2]
```diff
@@ -2891,27 +2891,43 @@ public void getBlobName(String inputName, String expectedOutputName) {
...
+
+    @ParameterizedTest
+    @MethodSource("getNonEncodedBlobNameSupplier")
+    public void getNonEncodedBlobName(String originalBlobName) {
+        BlobClient client = cc.getBlobClient(originalBlobName);
+        BlockBlobClient blockBlobClient = cc.getBlobClient(client.getBlobName()).getBlockBlobClient();
+        assertEquals(blockBlobClient.getBlobName(), originalBlobName);
+
+        // now try uploading to the portal and see if the blob name will be properly encoded in the url
+        blockBlobClient.upload(DATA.getDefaultInputStream(), DATA.getDefaultDataSize());
```
Comment (seanmcc-msft) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobApiTests.java:2928]: I don't think we actually need to do an upload here, we just care about the assert on 2930.

**Review by seanmcc-msft — COMMENTED** [review]
I think this change also affects Data Lake and Files

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeDirectoryAsyncClient.java [hunk 3]
```diff
@@ -1302,6 +1303,7 @@ SpecializedBlobClientBuilder prepareBuilderAppendPath(String pathName) {
         return new SpecializedBlobClientBuilder()
             .pipeline(getHttpPipeline())
             .serviceVersion(TransformUtils.toBlobServiceVersion(getServiceVersion()))
-            .endpoint(StorageImplUtils.appendToUrlPath(blobUrl, pathName).toString());
+            .endpoint(blobUrl)
+            .blobName(pathName);
```
[the remark this answers is not in the document]
Comment (ibrahimrabab) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeDirectoryAsyncClient.java:1305]: added subdirectoryAndFilePaths test in DirectoryApiTests.java, let me know if this is sufficient enough!

**Review by alzimmermsft — APPROVED** [review]
Approving with a thought.

Do we want to introduce an environment configuration that would re-enable the old behavior to give a fail-safe if code was depending on the behavior?

**Pull request merged**